src/NHibernate/Tuple/DynamicEntityInstantiator.cs

```diff
@@ -0,0 +1,64 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using NHibernate.Mapping;
+
+namespace NHibernate.Tuple
+{
+	[Serializable]
+	public class DynamicEntityInstantiator : IInstantiator
+	{
+		public const string Key = "$type$";
+
+		private readonly string _entityName;
+		private readonly HashSet<string> _isInstanceEntityNames = new HashSet<string>();
+
+		public DynamicEntityInstantiator(PersistentClass mappingInfo)
+		{
+			_entityName = mappingInfo.EntityName;
+			_isInstanceEntityNames.Add(_entityName);
+			if (mappingInfo.HasSubclasses)
+			{
+				foreach (var subclassInfo in mappingInfo.SubclassClosureIterator)
+					_isInstanceEntityNames.Add(subclassInfo.EntityName);
+			}
+		}
+
+		protected virtual IDictionary<string, object> GenerateMap()
+		{
+			return new Dictionary<string, object>();
+		}
+
+		#region IInstantiator Members
+
+		public object Instantiate(object id)
+		{
+			return Instantiate();
+		}
+
+		public object Instantiate()
+		{
+			var map = GenerateMap();
+			if (_entityName != null)
```

Comment:
This is always `true`.

Reply:
All NHibernate paths always assign a non-null `EntityName` to `PersistentClass`, but `PersistentClass` does not guarantee it will never be `null`. Now for the purpose here, we do not need a special handling of `null`. If `EntityName` is `null`, the old code would have added it to `_isInstanceEntityNames` nonetheless.